### PR TITLE
AM-199: dynamic creation of appointment blocks and time slots

### PR DESCRIPTION
**Pull request by jecihjoy**
PR for Ticket https://issues.openmrs.org/browse/AM-199
Kindly Review

### CODE REVIEW

**Comment by mogoodrich** [comment]
Thanks @jecihjoy !  This is a big one, so I might not get to it today, but will make sure I review it in the next day or two!

**Comment by mogoodrich** [comment]
@jecihjoy  I started looking, but I haven't finished yet... but one thing I noticed... can you add some tests?

**Comment by jecihjoy** [comment]
Yes @mogoodrich i'll be adding test.
Also check the changes I have made to the appointments payload validation, I have moved some checks to the service layer..

File: api/src/main/java/org/openmrs/module/appointmentscheduling/api/db/hibernate/HibernateAppointmentResourceDAO.java [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+                if (endDate != null)
+                    stringQuery += " AND appointmentResource.endDate=:endDate";
+                if (startTime != null)
+                    stringQuery += " AND appointmentResource.startTime=:startTime";
+                if (endTime != null)
+                    stringQuery += " AND appointmentResource.endTime=:endTime";
+                if (location != null)
+                    stringQuery += " AND appointmentResource.location=:location";
+                if (appointmentDate != null) {
+                    if (!getTime(appointmentDate).equals("00:00:00")) {
```
Comment (mogoodrich) [inline api/src/main/java/org/openmrs/module/appointmentscheduling/api/db/hibernate/HibernateAppointmentResourceDAO.java]: I'm confused if you pass in an appointment date, it only actually uses it in the if there is a time component?  What is the use case here?

Also, we shouldn't be doing a string comparison here to test that the time component is 0... you should be able to use the Joda Time library to do this.

File: api/src/main/java/org/openmrs/module/appointmentscheduling/api/db/hibernate/HibernateTimeSlotDAO.java [hunk 2]
```diff
@@ -72,5 +77,55 @@ else if (fromDate != null && toDate != null && !fromDate.before(toDate)) {
 		return super.sessionFactory.getCurrentSession().createCriteria(TimeSlot.class)
 		        .add(Restrictions.eq("appointmentBlock", appointmentBlock)).add(Restrictions.eq("voided", false)).list();
 	}
+
+	@Override
+	public TimeSlot getResourceTimeslot(Date date, Location location) throws DAOException {
```
Comment (mogoodrich) [inline api/src/main/java/org/openmrs/module/appointmentscheduling/api/db/hibernate/HibernateTimeSlotDAO.java]: Per my comment above, can you give a better description of what this method is intended to do?

File: api/src/main/java/org/openmrs/module/appointmentscheduling/api/impl/AppointmentServiceImpl.java [hunk 3]
```diff
@@ -1153,8 +1168,166 @@ public Appointment bookAppointment(Appointment appointment,
...
+
+			if (!allowOverbook) {
+				if (getTimeLeftInTimeSlot(appointment.getTimeSlot()) < appointment
+						.getAppointmentType().getDuration()) {
+					throw new TimeSlotFullException();
+				}
+			}
+
+			appointment.setStatus(AppointmentStatus.SCHEDULED);
+			return Context.getService(AppointmentService.class).saveAppointment(
```
Comment (mogoodrich) [inline api/src/main/java/org/openmrs/module/appointmentscheduling/api/impl/AppointmentServiceImpl.java]: Overall, can you give an explanation of how this method is supposed to work?

File: api/src/main/java/org/openmrs/module/appointmentscheduling/api/impl/AppointmentServiceImpl.java [hunk 4]
```diff
@@ -1153,8 +1168,166 @@ public Appointment bookAppointment(Appointment appointment,
...
+	}
+
+	@Override
+	public AppointmentResource getAppointmentResourceByUuid(String uuid) {
+		return (AppointmentResource) getAppointmentResourceDAO().getByUuid(uuid);
+	}
+
+	@Override
+	public AppointmentResource saveAppointmentResource(AppointmentResource appointmentResource) throws APIException {
+		appointmentResource.setIncludeWeekends(true);
```
Comment (mogoodrich) [inline api/src/main/java/org/openmrs/module/appointmentscheduling/api/impl/AppointmentServiceImpl.java]: Why do we always set IncludeWeekends to true?

File: api/src/main/java/org/openmrs/module/appointmentscheduling/api/impl/AppointmentServiceImpl.java [hunk 5]
```diff
@@ -1153,8 +1168,166 @@ public Appointment bookAppointment(Appointment appointment,
...
+	}
+
+	@Override
+	public BlockExcludedDays geyExcludedDayByUuid(String uuid) throws APIException {
+		return appointmentResourceDAO.geyExcludedDayByUuid(uuid);
+	}
+
+	private synchronized AppointmentResource saveAppointResourceInternal(AppointmentResource appointmentResource) throws APIException {
+		//check if resource already exists
+		List<AppointmentResource> resource = getAppointmentResourceDAO().getResourceByConstraints(appointmentResource.getStartDate(), appointmentResource.getEndDate(),
```
Comment (mogoodrich) [inline api/src/main/java/org/openmrs/module/appointmentscheduling/api/impl/AppointmentServiceImpl.java]: Could this test happen in the validator?

File: api/src/main/java/org/openmrs/module/appointmentscheduling/api/impl/AppointmentServiceImpl.java [hunk 6]
```diff
@@ -1153,8 +1168,166 @@ public Appointment bookAppointment(Appointment appointment,
...
+				appointmentResource.getStartTime(), appointmentResource.getEndTime(), appointmentResource.getLocation(), null);
+		if (resource.size() > 0) {
+			throw new APIException("Resource with those details already exist");
+		} else {
+			if (CollectionUtils.isEmpty(appointmentResource.getDaysList())) {
+				return (AppointmentResource) getAppointmentResourceDAO().saveOrUpdate(appointmentResource);
+			}
+			appointmentResource.setDaysList(null);
```
Comment (mogoodrich) [inline api/src/main/java/org/openmrs/module/appointmentscheduling/api/impl/AppointmentServiceImpl.java]: We are always setting the days list to null upon saving?  This seems confusing, or am I missing something?

File: api/src/main/java/org/openmrs/module/appointmentscheduling/api/impl/AppointmentServiceImpl.java [hunk 7]
```diff
@@ -1153,8 +1168,166 @@ public Appointment bookAppointment(Appointment appointment,
...
+
+			return (AppointmentResource) getAppointmentResourceDAO().saveOrUpdate(appointmentResource);
+		}
+
+	}
+
+
+	public TimeSlot createTimeSlot(Date date, Location location) {
+
+		List<AppointmentResource> ar = getAppointmentResourceDAO().getResourceByConstraints(null, null, null, null, location, date);
```
Comment (mogoodrich) [inline api/src/main/java/org/openmrs/module/appointmentscheduling/api/impl/AppointmentServiceImpl.java]: Doesn't this need (at least optionally) to take provider into account?

File: api/src/main/java/org/openmrs/module/appointmentscheduling/api/impl/AppointmentServiceImpl.java [hunk 8]
```diff
@@ -1153,8 +1168,166 @@ public Appointment bookAppointment(Appointment appointment,
...
+		if (ar != null) {
+			TimeSlot timeSlot = new TimeSlot();
+			for (AppointmentResource appointmentResource : ar) {
```
Comment (mogoodrich) [inline api/src/main/java/org/openmrs/module/appointmentscheduling/api/impl/AppointmentServiceImpl.java]: I'm a little confused, looks like you loop through all the possible resources here, and within the loop you create a appointment block, so you'd be creating a block-per-resource, but then the Time Slot is only saved when you are outside the loop?  So the time slot would only be for the last appointment block.

I would think you would just pick the "most relevant" appointment resource and create a single block and time slot?  (We'd need to determine how the "most relevant" would be determined)

File: api/src/main/java/org/openmrs/module/appointmentscheduling/validator/AppointmentValidator.java [hunk 9]
```diff
@@ -54,18 +54,18 @@ public void validate(Object obj, Errors errors) {
 		if (appointment == null) {
 			errors.rejectValue("appointment", "error.general");
 		} else {
-			ValidationUtils.rejectIfEmpty(errors, "timeSlot", "appointmentscheduling.Appointment.emptyTimeSlot");
+//			ValidationUtils.rejectIfEmpty(errors, "timeSlot", "appointmentscheduling.Appointment.emptyTimeSlot");
```
Comment (mogoodrich) [inline api/src/main/java/org/openmrs/module/appointmentscheduling/validator/AppointmentValidator.java]: why are we removing this validation?

File: api/src/main/java/org/openmrs/module/appointmentscheduling/validator/AppointmentValidator.java [hunk 10]
```diff
@@ -54,18 +54,18 @@ public void validate(Object obj, Errors errors) {
...
 			errors.rejectValue("appointment", "error.general");
 		} else {
-			ValidationUtils.rejectIfEmpty(errors, "timeSlot", "appointmentscheduling.Appointment.emptyTimeSlot");
+//			ValidationUtils.rejectIfEmpty(errors, "timeSlot", "appointmentscheduling.Appointment.emptyTimeSlot");
 			ValidationUtils.rejectIfEmpty(errors, "patient", "appointmentscheduling.Appointment.emptyPatient");
 			ValidationUtils.rejectIfEmpty(errors, "appointmentType", "appointmentscheduling.Appointment.emptyType");
 			
 			//Check whether the slot supports this appointment type
-			AppointmentType type = appointment.getAppointmentType();
+			/*AppointmentType type = appointment.getAppointmentType();
```
Comment (mogoodrich) [inline api/src/main/java/org/openmrs/module/appointmentscheduling/validator/AppointmentValidator.java]: why are we removing this validation?

File: api/src/main/java/org/openmrs/module/appointmentscheduling/BlockExcludedDays.java [hunk 11]
```diff
@@ -0,0 +1,58 @@
+package org.openmrs.module.appointmentscheduling;
+
+import org.openmrs.BaseOpenmrsData;
+
+import java.util.Date;
+
+public class BlockExcludedDays extends BaseOpenmrsData {
```
Comment (mogoodrich) [inline api/src/main/java/org/openmrs/module/appointmentscheduling/BlockExcludedDays.java]: Wouldn't this really be ResourceExcludedDays?

File: api/src/main/java/org/openmrs/module/appointmentscheduling/Appointment.java [hunk 12]
```diff
@@ -124,6 +126,10 @@ public String toString() {
 
 	private AppointmentType appointmentType;
 
+	private Date date;
+
+	private Location location;
```
[the remark this answers is not in the document]
Comment (jecihjoy) [reply api/src/main/java/org/openmrs/module/appointmentscheduling/Appointment.java]: With dynamic blocks feature, the appnt timeslot is set to the appnt object the the service. I'm using appnt date and appnt location to retrieve this timeslot so I require client to provide these two more properties in the appnt payload, that's why i'm adding them to the appnt domain.
My new payload will look like:
	"patient": "b7009090-4015-11e4-8e9b-2939a1809c7e",
	"status": "SCHEDULED",
	"appointmentType": "9ebdc232-c436-11e4-a470-82b0ea87e2d8",
	"date":"2020-04-18 17:00:00",
	"location":"b1a8b05e-3542-4037-bbd3-998ee9c40574"
excluding the timeslot

File: api/src/main/java/org/openmrs/module/appointmentscheduling/AppointmentResource.java [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+
+    private Time endTime;
+
+    private Provider provider;
+
+    private Location location;
+
+    private Set<AppointmentType> types;
+
+    private List<BlockExcludedDays> daysList;
```
[the remark this answers is not in the document]
Comment (jecihjoy) [reply api/src/main/java/org/openmrs/module/appointmentscheduling/AppointmentResource.java]: okay

File: api/src/main/java/org/openmrs/module/appointmentscheduling/api/db/hibernate/HibernateAppointmentResourceDAO.java (hunk 1 above)
Comment (jecihjoy) [reply api/src/main/java/org/openmrs/module/appointmentscheduling/api/db/hibernate/HibernateAppointmentResourceDAO.java]: The challenge i'm facing is some of these dates were actually created with type util date. Does it make sense when I change their date types to util date?

File: api/src/main/java/org/openmrs/module/appointmentscheduling/Appointment.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (jecihjoy) [reply api/src/main/java/org/openmrs/module/appointmentscheduling/Appointment.java]: @mogoodrich  **how do yu suggest I should **handle these transient/utility properties** 
Should I use a Data Transfer Object?**

**Pull request closed without merging**